### PR TITLE
offline: expose ability to get pending state without closing container

### PR DESCRIPTION
**Pull request by wes-carlson**
Fix [AB#4181](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/4181)

Expose existing ability to stash pending local changes from Container.closeAndGetPendingLocalState() but without closing the container.

This increases the risk of stash blob misuse. In addition to not reusing stash blobs, callers of this API must be careful not to use them after the container has reconnected, i.e. when it emits a "connected" event.

### CODE REVIEW

File: packages/common/container-definitions/src/loader.ts [hunk 1]
```diff
@@ -369,10 +369,20 @@ export interface IContainer extends IEventProvider<IContainerEvents>, IFluidRout
...
 
+	/**
+	 * Get pending state from container. WARNING: misuse of this API can result in duplicate op
+	 * submission and potential document corruption. The blob returned MUST be deleted if and when this
+	 * container emits a "connected" event.
+	 * @returns serialized blob that can be passed to Loader.resolve()
+	 * @experimental misuse of this API can result in duplicate op submission and potential document corruption
+	 * {@link https://github.com/microsoft/FluidFramework/blob/main/packages/loader/container-loader/closeAndGetPendingLocalState.md}
+	 */
+	getPendingLocalState(): string;
```
Comment (anthony-murphy) [inline packages/common/container-definitions/src/loader.ts:384]: in general i'd like to move away from adding experimental features to existing interfaces. it makes it difficult to manage breaking changes, and doesn't require distinct awareness from the consumer to take a dependency on such experimental features, specifically intellisense will show this by default, making it easy to accidentally use. 


For this PR i can see a few options:
- create new experimental interface in container loader package which requires and explicit cast from IContainer to something like IContainerExperimentalOffline to get access to the method
- in container loader only expose a function, getPendingLocalStateExperimental(container: IContainer) which internally cast the provide IContainer to Container to call the real methods.

File: packages/common/container-definitions/src/loader.ts [hunk 2]
```diff
@@ -657,3 +657,19 @@ export interface ISnapshotTreeWithBlobContents extends ISnapshotTree {
 	blobsContents: { [path: string]: ArrayBufferLike };
 	trees: { [path: string]: ISnapshotTreeWithBlobContents };
 }
+
+/**
+ * IContainer interface that includes experimental features still under development.
+ * @internal
+ */
+export interface IContainerExperimental extends IContainer {
```
Comment (anthony-murphy) [inline packages/common/container-definitions/src/loader.ts:665]: can you move this to the container-loader package

File: packages/loader/container-loader/src/container.ts [hunk 3]
```diff
@@ -997,6 +997,12 @@ export class Container
 		// runtime matches pending ops to successful ones by clientId and client seq num, so we need to close the
 		// container at the same time we get pending state, otherwise this container could reconnect and resubmit with
 		// a new clientId and a future container using stale pending state without the new clientId would resubmit them
+		const pendingState = this.getPendingLocalState();
+		this.close();
+		return pendingState;
+	}
+
+	public getPendingLocalState(): string {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:1005]: we should add the experimental interface to this class, Container to ensure its implemented correctly.

**Review by anthony-murphy — APPROVED** [review]

**Comment by vladsud** [comment]
Can we do anything to detect misuse faster? It's already easy to misuse it, this opens even more floodgate.
I'd strongly recommend putting experimental tag on getPendingLocalState() and a big warning on API itself why it's dangerous. Someone not knowing much about internal implementation details could start using it, without understanding danger.

**Comment by anthony-murphy** [comment]
> Can we do anything to detect misuse faster? It's already easy to misuse it, this opens even more floodgate. I'd strongly recommend putting experimental tag on getPendingLocalState() and a big warning on API itself why it's dangerous. Someone not knowing much about internal implementation details could start using it, without understanding danger.

The current usage requires turning on a feature flag, and casting the container to IContainerExperimental, and only then is the `getPendingLocalState` available. The method in the interface has a big scary comment warning.

We should additionally give the method a scarier name, like EXPERIMENTAL_getPendingLocalState? but i'm not sure what else we could do?

**Pull request merged**